catalyst/data/sampler.py

```diff
@@ -192,6 +193,146 @@ def __iter__(self) -> Iterator[int]:
         return iter(inds)
 
 
+class DynamicBalanceClassSampler(Sampler):
+    """
+    This kind of sampler can be used for classification tasks with significant
+    class imbalance.
+
+    The idea of this sampler that we start with the original class distribution
+    and gradually move to uniform class distribution like with downsampling.
+
+    Let's define D_i = #C_i/ #C_min where #C_i is a size of class i and #C_min
+    is a size of the rarest class, so D_i define class distribution.
+    Also define g(n_epoch) is a exponential scheduler. On each epoch
+    current D_i  calculated as current D_i  = D_i ^ g(n_epoch),
+    after this data samples according this distribution.
+
+    Note: In the end of the training, epochs will contain only
+    min_size_class * n_classes examples. So, possible it will not necessary to
+    do validation on each epoch. For this reason use ControlFlowCallback.
+
+    Usage example:
+
+        >>> import torch
+        >>> import numpy as np
+
+        >>> from catalyst.data.sampler import DynamicBalanceClassSampler
+        >>> from torch.utils import data
+
+        >>> features = torch.Tensor(np.random.random((200, 100)))
+        >>> labels = np.random.randint(0, 4, size=(200,))
+        >>> sampler = DynamicBalanceClassSampler(labels)
+        >>> labels = torch.LongTensor(labels)
+        >>> dataset = data.TensorDataset(features, labels)
+        >>> loader = data.dataloader.DataLoader(dataset, batch_size=8)
+
+        >>> for batch in loader:
+        >>>     b_features, b_labels = batch
+
+    Sampler was inspired by https://arxiv.org/pdf/1901.06783.pdf
+    """
+
+    def __init__(
+        self,
+        labels: List[Union[int, str]],
+        exp_lambda=0.9,
+        epoch: int = 0,
+        max_d: int = None,
+        mode: int = None,
+        ignore_warning: bool = False,
+    ):
+        """
+        Args:
+            labels: list of labels for each elem in the dataset
+            exp_lambda: exponent figure for schedule
+            epoch: start epoch number can be useful for many stage experiments
+            max_d: if not None, limit on the difference between the most
+            frequent and the rarest classes, heuristic
+            mode: if not None, it means the final class size in training.
```

Comment:
does this mean `the number of samples per class in the end`? why do we call it `mode that? or could we make in `Union[str, int]`, so it could take values of "upsampling", "downsampling", or some specified number of samples?

Reply:
I have added "downsampling". "Upsampling" doesn't work clearly